File: src/main/java/org/apache/fineract/client/services/TaxComponentsApi.java
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership. The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.fineract.client.services;

import org.apache.fineract.client.CollectionFormats.*;

import io.reactivex.Observable;
import io.reactivex.Completable;
import retrofit2.http.*;

import okhttp3.RequestBody;
import okhttp3.ResponseBody;
import okhttp3.MultipartBody;

import org.apache.fineract.client.models.GetTaxesComponentsResponse;
import org.apache.fineract.client.models.PostTaxesComponentsRequest;
import org.apache.fineract.client.models.PostTaxesComponentsResponse;
import org.apache.fineract.client.models.PutTaxesComponentsTaxComponentIdRequest;
import org.apache.fineract.client.models.PutTaxesComponentsTaxComponentIdResponse;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public interface TaxComponentsApi {
  /**
   * Create a new Tax Component
   * Creates a new Tax Component  Mandatory Fields: name, percentage  Optional Fields: debitAccountType, debitAcountId, creditAccountType, creditAcountId, startDate
   * @param postTaxesComponentsRequest  (required)
   * @return Observable&lt;PostTaxesComponentsResponse&gt;
   */
  @Headers({
    "Content-Type:application/json"
  })
  @POST("taxes/component")
  Observable<PostTaxesComponentsResponse> createTaxCompoent(
    @retrofit2.http.Body PostTaxesComponentsRequest postTaxesComponentsRequest
  );

  /**
   * List Tax Components
   * List Tax Components
   * @return Observable&lt;List&lt;GetTaxesComponentsResponse&gt;&gt;
   */
  @GET("taxes/component")
  Observable<List<GetTaxesComponentsResponse>> retrieveAllTaxComponents();
    

  /**
   * Retrieve Tax Component
   * Retrieve Tax Component
   * @param taxComponentId taxComponentId (required)
   * @return Observable&lt;GetTaxesComponentsResponse&gt;
   */
  @GET("taxes/component/{taxComponentId}")
  Observable<GetTaxesComponentsResponse> retrieveTaxComponent(
    @retrofit2.http.Path("taxComponentId") Long taxComponentId
  );

  /**
   * 
   * 
   * @return Observable&lt;String&gt;
   */
  @GET("taxes/component/template")
  Observable<String> retrieveTemplate20();
    

  /**
   * Update Tax Component
   * Updates Tax component. Debit and credit account details cannot be modified. All the future tax components would be replaced with the new percentage.
   * @param taxComponentId taxComponentId (required)
   * @param putTaxesComponentsTaxComponentIdRequest  (required)
   * @return Observable&lt;PutTaxesComponentsTaxComponentIdResponse&gt;
   */
  @Headers({
    "Content-Type:application/json"
  })
  @PUT("taxes/component/{taxComponentId}")
  Observable<PutTaxesComponentsTaxComponentIdResponse> updateTaxCompoent(
    @retrofit2.http.Path("taxComponentId") Long taxComponentId, @retrofit2.http.Body PutTaxesComponentsTaxComponentIdRequest putTaxesComponentsTaxComponentIdRequest
  );

}

File: src/main/kotlin/org/openapitools/client/apis/TaxComponentsApi.kt
package org.openapitools.client.apis

import org.openapitools.client.models.GetTaxesComponentsResponse
import org.openapitools.client.models.PostTaxesComponentsRequest
import org.openapitools.client.models.PostTaxesComponentsResponse
import org.openapitools.client.models.PutTaxesComponentsTaxComponentIdRequest
import org.openapitools.client.models.PutTaxesComponentsTaxComponentIdResponse
import retrofit2.http.Body
import retrofit2.http.GET
import retrofit2.http.POST
import retrofit2.http.PUT
import retrofit2.http.Path

interface TaxComponentsApi {
    /**
     * Create a new Tax Component
     * Creates a new Tax Component  Mandatory Fields: name, percentage  Optional Fields: debitAccountType, debitAcountId, creditAccountType, creditAcountId, startDate
     * Responses:
     *  - 200: OK
     *
     * @param postTaxesComponentsRequest
     * @return [PostTaxesComponentsResponse]
     */
    @POST("v1/taxes/component")
    suspend fun createTaxCompoent(@Body postTaxesComponentsRequest: PostTaxesComponentsRequest): PostTaxesComponentsResponse

    /**
     * List Tax Components
     * List Tax Components
     * Responses:
     *  - 200: OK
     *
     * @return [kotlin.collections.List<GetTaxesComponentsResponse]
     */
    @GET("v1/taxes/component")
    suspend fun retrieveAllTaxComponents(): List<GetTaxesComponentsResponse>

    /**
     * Retrieve Tax Component
     * Retrieve Tax Component
     * Responses:
     *  - 200: OK
     *
     * @param taxComponentId taxComponentId
     * @return [GetTaxesComponentsResponse]
     */
    @GET("v1/taxes/component/{taxComponentId}")
    suspend fun retrieveTaxComponent(@Path("taxComponentId") taxComponentId: Long): GetTaxesComponentsResponse

    /**
     *
     *
     * Responses:
     *  - 0: default response
     *
     * @return [kotlin.String]
     */
    @GET("v1/taxes/component/template")
    suspend fun retrieveTemplate21(): String

    /**
     * Update Tax Component
     * Updates Tax component. Debit and credit account details cannot be modified. All the future tax components would be replaced with the new percentage.
     * Responses:
     *  - 200: OK
     *
     * @param taxComponentId taxComponentId
     * @param putTaxesComponentsTaxComponentIdRequest
     * @return [PutTaxesComponentsTaxComponentIdResponse]
     */
    @PUT("v1/taxes/component/{taxComponentId}")
    suspend fun updateTaxCompoent(
        @Path("taxComponentId") taxComponentId: Long,
        @Body putTaxesComponentsTaxComponentIdRequest: PutTaxesComponentsTaxComponentIdRequest
    ): PutTaxesComponentsTaxComponentIdResponse

}
